metagraph/src/annotation/taxonomy/taxonomic_db.cpp

```diff
@@ -0,0 +1,326 @@
+    uint64_t num_taxid_failed = 0;
+    while (relevant_taxids.size()) {
+        const TaxId taxid = relevant_taxids.front();
+        relevant_taxids.pop();
+        if (normalized_taxid.count(taxid)) {
+            continue;
+        }
+        if (! full_parents_list.count(taxid)) {
+            logger->warn("Taxid {} cannot be found in the taxonomic tree.", taxid);
+            num_taxid_failed += 1;
+            continue;
+        }
+        normalized_taxid[taxid] = num_nodes++;
+        denormalized_taxid.push_back(taxid);
+        if (taxid == full_parents_list[taxid]) {
+            root_node = normalized_taxid[taxid];
+            continue;
+        }
+        relevant_taxids.push(full_parents_list[taxid]);
+    }
+    if (num_taxid_failed) {
+        logger->warn("Could not find {} taxid out of the total number of {} evaluated taxids.", num_taxid_failed, num_nodes);
+    }
+
+    tree.resize(num_nodes);
+    for (const auto &it: normalized_taxid) {
+        TaxId taxid = it.first;
+        if (normalized_taxid[taxid] == root_node) {
+            continue;
+        }
+        tree[normalized_taxid[full_parents_list[taxid]]].push_back(
+                normalized_taxid[taxid]);
+    }
+}
+
+void TaxonomyDB::rmq_preprocessing(const std::vector<NormalizedTaxId> &tree_linearization) {
+    uint num_rmq_rows = log2(tree_linearization.size()) + 1;
+    rmq_data.resize(num_rmq_rows);
+    for (uint64_t i = 0; i < num_rmq_rows; ++i) {
+        rmq_data[i].resize(tree_linearization.size());
+    }
+
+    // Copy tree_linearization on rmq[0].
+    for (uint64_t i = 0; i < tree_linearization.size(); ++i) {
+        rmq_data[0][i] = tree_linearization[i];
+    }
+
+    uint delta = 1;
+    for (uint row = 1; row < num_rmq_rows; ++row) {
+        for (uint64_t i = 0; i + delta < tree_linearization.size(); ++i) {
+            if (node_depth[rmq_data[row - 1][i]] > node_depth[rmq_data[row - 1][i + delta]]) {
+                rmq_data[row][i] = rmq_data[row - 1][i];
+            } else {
+                rmq_data[row][i] = rmq_data[row - 1][i + delta];
+            }
+        }
+        delta *= 2;
+    }
+
+    precalc_log2.resize(tree_linearization.size());
+    precalc_pow2.push_back(1);
+    for (uint64_t i = 2; i < tree_linearization.size(); ++i) {
+        precalc_log2[i] = 1 + precalc_log2[i/2];
+        if(precalc_log2[i] > precalc_log2[i-1]) {
+            precalc_pow2.push_back(i);
+        }
+    }
+}
+
+std::string TaxonomyDB::get_accession_version_from_label(const std::string &label) {
+    return  utils::split_string(label, "|")[3];
+}
+
+// TODO improve this by parsing the compressed ".gz" version (or use https://github.com/pmenzel/taxonomy-tools)
+void TaxonomyDB::read_lookup_table(const std::string &lookup_table_filepath,
+                                   const tsl::hopscotch_set<AccessionVersion> &input_accessions) {
+    //   TODO Print an error when the file format (e.g. '\t') is broken.
```

Comment:
I suggest implementing this now. Such error messages are often very helpful.

Reply:
I added some error messages. What do you think?